### PR TITLE
Twitter provider

### LINKED ISSUE #4: Twitter provider
Are you interested in this? Would you accept such a pull request?

The problem is twitter uses OAuth 1.0 (not 2.0), but this does not violate shard name. But will need to change the interfaces though. 

https://dev.twitter.com/oauth

### PR DESCRIPTION
**Pull request by veelenga**
Implementation of Twitter provider OAuth 1.0. The flow is described [here](https://dev.twitter.com/web/sign-in/implementing).

This change also introduces different providers for OAuth and OAuth2 protocols, `OAuthProvider` and `OAuth2Provider` respectively. The [difference](https://github.com/veelenga/multi_auth/blob/935d2b70ac4a10fa9acbc123e5588e14389799ac/src/multi_auth/provider.cr) between them is just in oauth params. 

## Bracking changes:

* `MultiAuth::Provider` now does not have attributes `redirect_uri`, `client_id`, `client_secret` since they are moved to `MultiAuth::OAuth2Provider`.

## Usage:

1. Configure:

```crystal
MultiAuth.config("twitter", consumer_key, consumer_secret)
```

2. Authorize url:

```crystal
MultiAuth.make("twitter", oauth_callback).authorize_url
# => https://api.twitter.com/oauth/authorize?oauth_token=token&oauth_callback=callback_url
```

screen shot 2017-09-25 at 12 01 59 pm

3. Obtain user attributes:

```crystal
# in the action routed by callback url
MultiAuth.make("twitter", callback_url).provider.user(params.to_h)
# => #<MultiAuth::User:0x10e584a00
#        @access_token=
#         #<OAuth::AccessToken:0x10e58ec30
#         @extra=
#           {"user_id" => "12313123",
#            "screen_name" => "test",
#           "x_auth_expires" => "0"},
#          @secret="XXX",
#          @token="XXX">,
#        @description=
#         "Creating software. #ruby #crystallang #elixirlang.",
#        @email="me@twitter.com",
#        @first_name=nil,
#        @image=
#         "http://pbs.twimg.com/profile_images/651454953739804672/-HpXVkau_normal.jpg",
#        @last_name=nil,
#        @location="Ukraine",
#        @name="V. Elenhaupt",
#        @nickname="veelenga",
#        @phone=nil,
#        @provider="twitter",
#       ...
```

closes #4

* https://oauth.net/core/1.0/
* https://stackoverflow.com/questions/4113934/how-is-oauth-2-different-from-oauth-1
* https://dev.twitter.com/web/sign-in/implementing
* https://dev.twitter.com/rest/reference/get/account/verify_credentials

### CODE REVIEW

**Comment by msa7** [comment]
Great work! Thank you!

File: src/multi_auth/provider.cr [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
-  getter redirect_uri : String
-  getter client_id : String
-  getter client_secret : String
+module MultiAuth
+  abstract class Provider
+    abstract def authorize_uri(scope = nil)
+    abstract def user(params : Hash(String, String))
+  end
+
+  abstract class OAuthProvider < Provider
```
Comment (msa7) [inline src/multi_auth/provider.cr]: The API of both `OAuthProvider`, `OAuth2Provider ` provide is same, is not it? Shell we combine them.

Comment (veelenga) [reply src/multi_auth/provider.cr]: Hm, maybe. But the API is not the same:

```
OAuthProvider#oauth_callback
OAuthProvider#consumer_key
OAuthProvider#consumer_secret

OAuth2Provider#redirect_uri
OAuth2Provider#client_id
OAuth2Provider#client_secret
```

I don't like the idea to combine those attributes into one class, so each OAuth2 provider will have redundant `comsumer_key` and `consumer_secret` attributes and vice versa...

Maybe it is possible to just do:

```
Provider#callback
Provider#key
Provider#secret
```

but that may lead to confusion with the required attributes to pass in...

What is your suggestion?

Comment (msa7) [reply src/multi_auth/provider.cr]: Personally, I would `just do`:

```
Provider#callback_uri
Provider#key
Provider#secret
```
All providers need three params, this described in the abstract provider.

**Comment by veelenga** [comment]
Thanks for the review and merge. May I ask you to create a new release?

**Pull request merged**